main.go

```diff
@@ -143,62 +166,28 @@ func getLightningIPsAndDataDirs() (lightningIPs, dataDirs []string, err error) {
 }
 
 /**
-> rm -rf /tmp/benchmarksql
-> git clone https://github.com/pingcap/benchmarksql.git /tmp/benchmarksql
-> yum install -y java ant
-> cd /tmp/benchmarksql
-> ant
-> sed -i 's/localhost:4000/tidb_ip:tidb_port' /tmp/benchmarksql/run/props.mysql
-> sed -i "s/warehouses=[0-9]\+/warehouses=10000/" /tmp/benchmarksql/run/props.mysql
-> echo fileLocation=%s >> /tmp/benchmarksql/run/props.mysql
-> echo tableName=%s >> /tmp/benchmarksql/run/props.mysql
+> rm -f /tmp/go-tpc
+> wget -O /tmp/go-tpc binary_url; chmod +x /tmp/go-tpc
 */
-func fetchTpccRepoAndEnforceConf(tidbIP, tidbPort string, warehouse int64, lightningDirs []string, lightningIPs []string) (err error) {
-	var tableName []string
-	switch len(lightningIPs) {
-	case 1:
-		tableName = []string{"all"}
-	case 2:
-		tableName = []string{"customer", "stock,order"}
-	case 3:
-		tableName = []string{"customer", "stock", "order"}
-	}
-
+func fetchTpcc(lightningDirs []string, lightningIPs []string, binaryURL string, skipDownloading bool) (err error) {
 	errCh := make(chan error, 3)
 	wg := &sync.WaitGroup{}
 	for i, lightningIP := range lightningIPs {
-		tn := tableName[i]
 		ip := lightningIP
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			if _, _, err = runCmd("ssh", ip, `rm -rf /tmp/benchmarksql`); err != nil {
-				errCh <- err
-				return
-			}
-			if _, _, err = runCmd("ssh", ip, `cd /tmp; git clone -b specify_table https://github.com/XuHuaiyu/benchmarksql.git`); err != nil {
-				errCh <- err
-				return
-			}
-			if _, _, err = runCmd("ssh", ip, `sudo yum install -y java ant`); err != nil {
-				errCh <- err
-				return
-			}
-			if _, _, err = runCmd("ssh", ip, fmt.Sprintf(`cd /tmp/benchmarksql; ant`)); err != nil {
-				errCh <- err
-				return
-			}
-			if _, _, err = runCmd("ssh", ip, fmt.Sprintf(`sed -i '%s' %s`, fmt.Sprintf(`s/localhost:4000/%s/;s/warehouses=[0-9]\+/%s/;s/loadWorkers=[0-9]\+/loadWorkers=%d/`, tidbIP+":"+tidbPort, fmt.Sprintf("warehouses=%d", warehouse), runtime.NumCPU()), "/tmp/benchmarksql/run/props.mysql")); err != nil {
-				errCh <- err
-				return
-			}
-			if _, _, err = runCmd("ssh", ip, fmt.Sprintf("echo fileLocation=%s/tpcc. >> /tmp/benchmarksql/run/props.mysql", lightningDirs[i])); err != nil {
-				errCh <- err
-				return
-			}
-			if _, _, err = runCmd("ssh", ip, fmt.Sprintf("echo tableName=%s >> /tmp/benchmarksql/run/props.mysql", tn)); err != nil {
-				errCh <- err
-				return
+			if !skipDownloading {
+				if _, _, err = runCmd("ssh", ip, `rm -f /tmp/go-tpc`); err != nil {
+					errCh <- err
+					return
+				}
+
+				if _, _, err = runCmd("ssh", ip, fmt.Sprintf("wget -O /tmp/go-tpc.tar.gz %s; tar -xvf /tmp/go-tpc.tar.gz -C /tmp/; rm -f /tmp/go-tpc.tar.gz; chmod +x /tmp/go-tpc", binaryURL)); err != nil {
+					errCh <- err
+					return
+				}
+				fmt.Println("Download go-tpc binary successfully!")
 			}
 			if _, _, err = runCmd("ssh", ip, fmt.Sprintf("mkdir -p %s", lightningDirs[i])); err != nil {
```

Comment:
loop variable i captured by func literal

Reply:
Good catch! Will update soon

Reply:
```go
-       threads   = flag.Int64(nmThreads, 40, "number of threads")
+       threads   = flag.Int64(nmThreads, 40, "number of threads of go-tpc")

        //ansibleDir = flag.String(nmAnsibleDir, "", "ansible directory path")

@@ -170,12 +170,11 @@ func getLightningIPsAndDataDirs() (lightningIPs, dataDirs []string, err error) {
 > wget -O /tmp/go-tpc binary_url; chmod +x /tmp/go-tpc
 */
 func fetchTpcc(lightningDirs []string, lightningIPs []string, binaryURL string, skipDownloading bool) (err error) {
-       errCh := make(chan error, 3)
+       errCh := make(chan error, len(lightningIPs)*10) // 3 should enough, echo run 3 cmd at most
        wg := &sync.WaitGroup{}
-       for i, lightningIP := range lightningIPs {
-               ip := lightningIP
+       for i := 0; i < len(lightningIPs); i++ {
                wg.Add(1)
-               go func() {
+               go func(ip string, dir string) {
                        defer wg.Done()
                        if !skipDownloading {
                                if _, _, err = runCmd("ssh", ip, `rm -f /tmp/go-tpc`); err != nil {
@@ -189,16 +188,18 @@ func fetchTpcc(lightningDirs []string, lightningIPs []string, binaryURL string,
                                }
                                fmt.Println("Download go-tpc binary successfully!")
                        }
-                       if _, _, err = runCmd("ssh", ip, fmt.Sprintf("mkdir -p %s", lightningDirs[i])); err != nil {
+                       if _, _, err = runCmd("ssh", ip, fmt.Sprintf("mkdir -p %s", dir)); err != nil {
                                errCh <- err
                                return
                        }
-               }()
+               }(lightningIPs[i], lightningDirs[i])
        }
+
        go func() {
                wg.Wait()
                close(errCh)
        }()
+
```
also looks maybe block at push into `errCh` if having multi lightning, have we test it with mutl lightning?
I make a change but has no right to push